### ISSUE TITLE
Support for Xamarin.iOS and Xamarin.Android

### ISSUE THREAD

**Issue by JamWils** [moq/moq4#102]
I've tried using Moq for unit testing on these platforms and seem unable to do so.  Has anyone checked whether this is feasible or not?  I think trying to move this to a PCL would help with this matter.  I can look into it myself if that is fine?  I'm new to helping out with an open source project.  I love using moq for testing and would like to use it on those platforms.

**Comment by machadogj** [comment]
Hi @JamWils, have you guys made a decision yet?

**Comment by JamWils** [comment]
@machadogj, unfortunately, I just don't have the time to look into this like I was hoping too.  If I ever get the chance I will let you know.  It would be great to have moq running on the clients of both iOS and Android.  I am spending most of my time on iOS Swift these days.  Thanks!

**Comment by jamesathey** [comment]
I've made a lot of progress toward [a working Xamarin.Android port of Castle.Core](https://github.com/jamesathey/Core/tree/xamarin-android). I've [submitted the first pull request](https://github.com/castleproject/Core/pull/116), but there will need to be two or three more after that before the work is fully upstreamed.

If moq on Xamarin.Android matters to you as well, let @jonorossi know.

**Comment by jonorossi** [comment]
There shouldn't be a need to use Cecil or something else as DynamicProxy on the .NET Framework allows you to persist the dynamic assembly it creates to disk. I recall some people doing this to run NHibernate with proxy support under medium trust deployments many years ago. Take a look at the unit tests (in the base class) and you'll see just about every unit test also does this to run peverify after generating code. This might be a solution for iOS.

**Comment by jamesathey** [comment]
FYI, I have completed my [moq4 port to Xamarin.Android](https://github.com/jamesathey/moq4/tree/android). It passes all of the unit tests using my android branch of Castle. Once that android branch is merged into Castle, I will submit a pull request here for the changes to Moq.

**Comment by kzu** [comment]
Sounds great!

**Comment by stakx** [comment]
@jamesathey, @JamWils, I would like to close this issue for two reasons:

1. I think it would be better to track support for Xamarin.iOS and Xamarin.Android separately. Judging from the above information, these two platforms have rather different support for dynamic type generation, and making Moq work on them will likely involve finding different solutions for each.

    May I suggest that, if you are still interested in support for those targets, you re-open a separate issue per specific target?

2. It seems that work on this has slowed, or even stalled; both here at the Moq repository, and over at Castle.Core's.

    @jamesathey, you appear to have been successful of getting Castle.Core to run on Xamarin.Android. If I interpret https://github.com/castleproject/Core/issues/135 correctly, it took some modifications based on Castle's .NET Standard 1.3 target? Are you still planning to send PRs towards Castle.Core to get them to support the Xamarin.Android target directly? That would probably be a precondition before more Moq can happen at Moq's end.

**Comment by karlingen** [comment]
Here's an alternative that should work on Xamarin.iOS, Xamarin.Android, Xamarin.UWP https://github.com/kentcb/PCLMock

**Comment by stakx** [comment]
Also, stay tuned for Moq 5, which also performs compile-time code generation instead of runtime dynamic proxying.

**Comment by karlingen** [comment]
@stakx Where can I subscribe to Moq 5 updates? :)

**Comment by stakx** [comment]
Subscribe to the https://github.com/moq/moq repo where v5 is being developed. You can also join Moq's Gitter chat at https://gitter.im/moq/moq.
